### ISSUE TITLE
Vertex does not handle configurable product with correlative simple product numeric sku

### ISSUE THREAD

**Issue by sebdelta** [magento/magento2#25471]
### Preconditions (*)
1. Magento 2.3.3
2. MariaDB 10.0.38

### Steps to reproduce (*)
1. Install Magento 2
2. Enable Vertex and set use_for_calculation=1
3. Create new sub-category
4. Add attribute color to product details attribute group of default attribute-set
5. Add 2 options to color attribute
6. Create product 1 with one of the color option and SKU `211890`
7. Create product 2 with other color option and a random SKU
8. Create configurable product with previous products as its children and SKU `211890c`
9. Go to front-end and go on the category page, select the configurable product, select the option associated with the product 1
10. Click add to cart

### Expected result (*)
1. Product should be successfully added to the cart

### Actual result (*)
1. Error message "We can't add this item to your shopping cart right now."
2. Exception.log message:
```
[2019-11-04 20:55:52] report.CRITICAL: Notice: Undefined index: 211890C in /vagrant/application2/mage2.vm/vendor/vertex/module-tax/Model/Plugin/CommonTaxCollectorPlugin.php on line 126 {"exception":"[object] (Exception(code: 0): Notice: Undefined index: 211890C in /vagrant/application2/mage2.vm/vendor/vertex/module-tax/Model/Plugin/CommonTaxCollectorPlugin.php on line 126 at /vagrant/application2/mage2.vm/vendor/magento/framework/App/ErrorHandler.php:61)"} []
```

**Comment by sebdelta** [comment]
@magento give me 2.3-develop instance

**Comment by magento-engcom-team** [comment]
Hi @sebdelta. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @sebdelta, here is your [Magento instance](https://i-25471-2-3-develop.instances.magento-community.engineering).
Admin access: [https://i-25471-2-3-develop.instances.magento-community.engineering/admin](https://i-25471-2-3-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by sebdelta** [comment]
@sdzhepa this issue is generated by the Vertex_Tax module. Is there a way to enable that module on the above vanilla Magento instance?

**Comment by sebdelta** [comment]
I tracked the issue to the file vendor/vertex/module-tax/Model/Plugin/CommonTaxCollectorPlugin.php of the vertex/module-tax module. The line 103 `$productCodes = array_keys($processItems);` retrieves the product skus, the problem is that PHP cast the numeric array keys to int (e.g. 211890), so $productCodes  can contain integer values which are being passed to the criteria for building the database query. The query will end up filtering a numeric value on the varchar sku columns. Mariadb will truncate with a warning the values in the database to match the numeric format (e.g. 211890c will become 211890 which will match) and multiple result for one sku can be returned which will throw an error later in the code trying to retrieve invalid skus in the $processItems array.
Attached is a proposed patch for vertex/module-tax  to cast all skus to string.
[MG-fix-vertex-skus-type-casting-patch.txt](https://github.com/magento/magento2/files/3806531/MG-fix-vertex-skus-type-casting-patch.txt)

**Comment by sdzhepa** [comment]
Hello @sebdelta 

It is not possible to enable Vertex_Tax module on vanilla Magento instance.
Instance deployer uses only git-based Magento installation, Vertex it is bundle extension included only to the composer-based installation. 


We have notified the vendor owner of Vertex about this report and results of your investigation from https://github.com/magento/magento2/issues/25471#issuecomment-549595019

Internal Jira ticket to track progress for this issue has been created `BUNDLE-2310`

We will provide updates as soon as we have it

**Comment by sebdelta** [comment]
@sdzhepa thank you for the update.

**Comment by navarr** [comment]
@sebdelta Thank you for your report.  We hope to have this fixed in 2.3.4

Fantastic catch of an interesting edgecase.  My apologies for the inconvenience.

**Comment by navarr** [comment]
This issue is resolved in `vertex/module-tax` version `3.3.0`.  This version is planned to be bundled with 2.3.4
